core/src/main/java/com/graphhopper/routing/querygraph/QueryGraph.java

```diff
@@ -592,110 +309,61 @@ public EdgeExplorer createEdgeExplorer(final EdgeFilter edgeFilter) {
         }
     }
 
-    private EdgeExplorer createUncachedEdgeExplorer(EdgeFilter edgeFilter) {
-        // Iteration over virtual nodes needs to be thread safe if done from different explorer
-        // so we need to create the mapping on EVERY call!
-        // This needs to be a HashMap (and cannot be an array) as we also need to tweak edges for some mainNodes!
-        // The more query points we have the more inefficient this map could be. Hmmh.
-        final IntObjectMap<VirtualEdgeIterator> node2EdgeMap
-                = new GHIntObjectHashMap<>(queryResults.size() * 3);
-
+    private EdgeExplorer createUncachedEdgeExplorer(final EdgeFilter edgeFilter) {
+        // build a map of virtual edge lists for every node that is modified compared to the mainGraph. this map
+        // depends on the given edgeFilter, but we could just as well build this map independent from the edge filter
+        // and apply the filter while iterating the edges
```

Comment:
I think applying the filter while iterating the edges would be cleaner, performance-wise I do not expect a difference ?

Reply:
What would be the advantage - less memory used?

Reply:
yes but that does not matter anyway. I dont know it just seems ugly to build this map beforehand for every edge filter. And maybe we can even get rid of query graph cache ?

Reply:
Yes, definitely ugly. The advantage is a huge boost to the performance for e.g. map matching and also matrix algorithms.

Reply:
> Yes, definitely ugly. The advantage is a huge boost to the performance for e.g. map matching and also matrix algorithms. (Basically every algo that requires many snaps/locations)

I dont know. The bottleneck for many snaps/locations is the LocationIndex lookup, not building the QueryGraph. The query graph cache I think is an improvement because we do not need to build the map for every edge explorer (saves memory but does it really speed up the request?). But now that we build most of the map beforehand I do not expect a performance issue if we apply the filter while iterating. But will try (better than thinking :))

Reply:
> The bottleneck for many snaps/locations is the LocationIndex lookup, not building the QueryGraph.

Ups, yes. In my head this is still tightly coupled :)

Reply:
Ok I have changed it such that QueryGraph first builds a map of all virtual edge states and virtual edge iterator filters the edges while calling next. Now this is as fast (if not faster) as master, even without the query graph cache. I think we can remove the query graph cache entirely (in a separate PR/commit).

I ran a little measurement with via routing for 5000 points for Germany (with and without query graph cache). With the cache only 5 edge explorers are created, without there are 24995, but since they are rather small objects and performance is no longer improved by the cache I dont think the added complexity of query graph cache is really worth it after this PR).

| branch | time routing | time query graph explorer building |
|-|-|-|
|master|270s|250s|
|master, cached|16s|50ms|
|query_graph_refactoring (this PR)|15s|2ms|
|query_graph_refactoring (this PR), cached|15s|1ms|

Reply:
Wow, this is rather unexpected and really nice of course :)

I fear I still do not understand the reason. What exactly was slow in uncached master? I always thought it was the EdgeExplorer creation :)

Reply:
> I dont know. The bottleneck for many snaps/locations is the LocationIndex lookup, not building the QueryGraph. The query graph cache I think is an improvement because we do not need to build the map for every edge explorer (saves memory but does it really speed up the request?).

I think looking at my little table I have to correct this: For many locations the edge explorer creation indeed was a bottleneck (and query graph cache fixed this). The interesting question here however is what 'edge explorer creation' means. In previous (uncached) master every time we called `createEdgeExplorer` we built a mapping between node ids and `VirtualEdgeIterator` for all virtual nodes and all real nodes adjacent to real nodes. (for 5000 locations this would be something like 15000). I think the only reason we had to rebuild this mapping everytime is because `VirtualEdgeIterator` stores the (mutable) `currentIndex`. What I changed in this PR is that I just build the mapping once at the time the `QueryGraph` is built, but instead of a map of `VirtualEdgeIterator` I create a map of (more or less immutable) `List<VirtualEdgeIterator**State**>`. 

> What exactly was slow in uncached master? I always thought it was the EdgeExplorer creation :)

Building the `IntObjectMap<VirtualEdgeIterator>`.

Another improvement I did is that I no longer put all virtual nodes and adjacent real nodes into the same mapping, but only the real nodes. The virtual nodes can be retrieved from a list, because all the virtual node ids are consecutive. This is a small improvement, because the hashmap stays a bit smaller.